modules/platform/nuxeo-apidoc-server/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/documentation/SecureXMLHelper.java

```diff
@@ -0,0 +1,209 @@
+import javax.xml.stream.events.XMLEvent;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.services.config.ConfigurationService;
+
+/**
+ * Helper for XML secure content management.
+ *
+ * @since 11.2
+ */
+public class SecureXMLHelper {
+
+    private static final Logger log = LogManager.getLogger(SecureXMLHelper.class);
+
+    protected static final String KEYWORDS_PROPERTY = "org.nuxeo.apidoc.secure.xml.keywords";
+
+    public static final List<String> DEFAULT_KEYWORDS = List.of("password", "Password", "secret", "apiKey");
+
+    protected static final String WHITELISTED_KEYWORDS_PROPERTY = "org.nuxeo.apidoc.secure.xml.keywords.whitelisted";
+
+    public static final List<String> DEFAULT_WHITELISTED_KEYWORDS = List.of("passwordField", "passwordHashAlgorithm");
+
+    protected static final String SECRET_VALUE = "********";
+
+    protected static final XMLInputFactory inputFactory = XMLInputFactory.newInstance();
+
+    protected static final XMLOutputFactory outputFactory = XMLOutputFactory.newInstance();
+
+    protected static final XMLEventFactory eventFactory = XMLEventFactory.newInstance();
+
+    /**
+     * Makes sure no passwords and similar sensitive data are embedded in the XML.
+     */
+    public static String secure(String xml) {
+        if (StringUtils.isBlank(xml)) {
+            return xml;
+        }
+        List<String> keywords = getKeywords();
+        List<String> whitelist = getWhitelistedKeywords();
+        try {
+            return secureStAX(xml, keywords, whitelist);
+        } catch (XMLStreamException e) {
+            log.error(e, e);
+            return secureRegexp(xml, keywords, whitelist);
+        }
+    }
+
+    public static List<String> getKeywords() {
+        return getKeywordList(KEYWORDS_PROPERTY, DEFAULT_KEYWORDS);
+    }
+
+    public static List<String> getWhitelistedKeywords() {
+        return getKeywordList(WHITELISTED_KEYWORDS_PROPERTY, DEFAULT_WHITELISTED_KEYWORDS);
+    }
+
+    protected static List<String> getKeywordList(String property, List<String> defaultValue) {
+        return Framework.getService(ConfigurationService.class)
+                        .getString(property)
+                        .map(v -> v.split("\\s*,[,\\s]*"))
+                        .map(List::of)
+                        .orElse(defaultValue);
+    }
+
+    public static String secureStAX(String xml, List<String> keywords, List<String> whitelist)
+            throws XMLStreamException {
+        // System.out.println("xml = " + xml);
+
+        InputStream stream = new ByteArrayInputStream(xml.getBytes());
+        XMLEventReader reader = inputFactory.createXMLEventReader(stream);
+        OutputStream output = new ByteArrayOutputStream();
+        XMLEventWriter writer = outputFactory.createXMLEventWriter(output);
```

Comment:
The `stream` InputStream doesn't seem to be closed, could we use try-with-resources to do that?

Reply:
As discussed, byte array input streams do not need to be closed, however could be a good practice, wdyt @efge please?

Reply:
@atchertchian indeed `ByteArrayInputStream` doesn't need to be closed, the GC will deal with it, however most tools and IDEs won't recognize that the `InputStream` you use later on comes from a `ByteArrayInputStream` and will flag it as a leaked resource. So you might as well use try-with-resources there too.